proposals/generic-ref-struct.md

```diff
@@ -0,0 +1,424 @@
+Unconstraining our generics
+=====
+
+## Summary
+This proposal is an aggregation of several different proposals for making `ref struct` more generally usable in our type system. The goal being to unify code paths today that have to specialize on `ref struct` and other types do to the inability of `ref struct` to participate in generics and implement interfaces. This proposal will allow us to generalize on `ref struct` as naturally as other types. 
+
+## Motivation
+This feature will allow developers to remove `ref struct` specific code by instead letting them implement interfaces that are used in generic contexts to generalize algorithms. For example an explicit goal is to allow `ref struct` to implement `ISpanFormattable`. The `ISpanFormattable` interface and `ref struct` are used in low level perf sensitive areas but our current language limitations prevent them from being used together.
+
+This will also allow for combinations like `Span<Span<T>>`. This is a long desire of the runtime team who often want to represent collections of `Span<T>` but are unable to do so due to our current limitations.
+
+It also will enable the creation of generic collections of `ref struct`. For instance having a `List<T>` style [data structure](#ref-list) with a hard capacity limit will be possible with the features in this proposal.
+
+## Detailed Design 
+
+### ref struct generic arguments
+The compiler will allow type parameters to be (un)constrained such that they can accept `ref struct` as arguments using the `where T : ref struct` syntax. 
+
+```c#
+void Swap<T>(ref T left, ref T right)
+    where T : ref struct
+{
+    T temp = left;
+    left = right;
+    right = temp;
+}
+
+Span<byte> span1 = new byte[4];
+Span<byte> span2 = default;
+
+// Can use ref struct type arguments 
+Swap<Span<byte>>(ref span1, ref span2);
+
+// Can also use normal type arguments
+Swap<byte>(ref span1[0], ref span1[1]);
+```
+
+At the call site this relaxes the set of legal type arguments to inclued `ref struct` *in addition* to the types already allowed. Even though this appears in the constraint position it serves more of an anti-constraint because it expands the set of possible types. 
```

Comment:
```suggestion
At the call site this relaxes the set of legal type arguments to included `ref struct` *in addition* to the types already allowed. Even though this appears in the constraint position it serves more of an anti-constraint because it expands the set of possible types. 
```

Reply:
```suggestion
At the use site this relaxes the set of legal type arguments to include `ref struct` *in addition* to the types already allowed. Even though this appears in the constraint position it serves more of an anti-constraint because it expands the set of possible types. 
```